runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+	config::assets::TrustBackedAssetsInstance, AccountId, AllowedApiCalls, Balance, Runtime,
+	RuntimeCall, RuntimeOrigin, UNIT,
+};
+use pop_primitives::{
+	cross_chain::CrossChainMessage,
+	nfts::{CollectionId, ItemId},
+	storage_keys::{
+		AssetsKeys::{self, *},
+		NftsKeys, ParachainSystemKeys, RuntimeStateKeys,
+	},
+	AssetId,
+};
+
+mod v0;
+
+const LOG_TARGET: &str = "pop-api::extension";
+// Versions:
+const V0: u8 = 0;
+
+type ContractSchedule<T> = <T as pallet_contracts::Config>::Schedule;
+
+#[derive(Default)]
+pub struct PopApiExtension;
+
+// TODO: check removal or simplification of trait bounds.
+impl<T> ChainExtension<T> for PopApiExtension
+where
+	T: pallet_contracts::Config
+		+ pallet_xcm::Config
+		+ pallet_assets::Config<TrustBackedAssetsInstance, AssetId = AssetId>
+		+ pallet_nfts::Config<CollectionId = CollectionId, ItemId = ItemId>
+		+ cumulus_pallet_parachain_system::Config
+		+ frame_system::Config<
+			RuntimeOrigin = RuntimeOrigin,
+			AccountId = AccountId,
+			RuntimeCall = RuntimeCall,
+		>,
+	T::AccountId: UncheckedFrom<T::Hash> + AsRef<[u8]>,
+{
+	fn call<E: Ext>(&mut self, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+	where
+		E: Ext<T = T>,
+	{
+		log::debug!(target:LOG_TARGET, " extension called ");
+		let mut env = env.buf_in_buf_out();
+		// Charge weight for making a call from a contract to the runtime.
+		// `debug_message` weight is a good approximation of the additional overhead of going
+		// from contract layer to substrate layer.
+		// reference: https://github.com/paritytech/ink-examples/blob/b8d2caa52cf4691e0ddd7c919e4462311deb5ad0/psp22-extension/runtime/psp22-extension-example.rs#L236
+		let contract_host_weight = ContractSchedule::<T>::get().host_fn_weights;
+		env.charge_weight(contract_host_weight.debug_message)?;
+
+		// Extract version and function_id from first two bytes.
+		let (version, function_id) = {
+			let bytes = env.func_id().to_le_bytes();
+			(bytes[0], bytes[1])
+		};
+		// Extract pallet index and call / key index from last two bytes.
+		let (pallet_index, call_index) = {
+			let bytes = env.ext_id().to_le_bytes();
+			(bytes[0], bytes[1])
+		};
+
+		let result = match FuncId::try_from(function_id) {
+			Ok(function_id) => {
+				// Read encoded parameters from buffer and calculate weight for reading `len` bytes`.
+				// reference: https://github.com/paritytech/polkadot-sdk/blob/117a9433dac88d5ac00c058c9b39c511d47749d2/substrate/frame/contracts/src/wasm/runtime.rs#L267
+				let len = env.in_len();
+				env.charge_weight(contract_host_weight.return_per_byte.saturating_mul(len.into()))?;
+				let params = env.read(len)?;
+				log::debug!(target: LOG_TARGET, "Read input successfully");
+				match function_id {
+					FuncId::Dispatch => {
+						dispatch::<T, E>(&mut env, version, pallet_index, call_index, params)
+					},
+					FuncId::ReadState => {
+						read_state::<T, E>(&mut env, version, pallet_index, call_index, params)
+					},
+					// TODO
```

Comment:
If SendXcm was not a `FuncId` it could be refactored to matching on the function_id after matching on the pallet index. If we don't need a `FuncId` for xcm related calls and match on its pallet index, we might not need the `FuncId` at all and have 256 chain extension functions per pallet. I think I see the problem now because this might not be enough for xcm use cases.

Reply:
The idea is to have predefined xcm programs, with parameters, to fulfil some use case, but this may be impractical when it comes to more advanced requirements.